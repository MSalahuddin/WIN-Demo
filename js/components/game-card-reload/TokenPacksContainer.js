import React, { useState, useContext, useEffect } from 'react';
import { StyleSheet, Platform } from 'react-native';
import PropTypes from 'prop-types';
import styled from 'styled-components/native';
import { KeyboardAwareFlatList, KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import branch from 'react-native-branch';
import * as RNIap from 'react-native-iap';
import TokenCard from './TokenCard';
import PromoCodeBox from './PromoCodeBox';
import PaymentMethodCard from './PaymentMethodCard';
import api from '../../api';
import Text, { SIZE, FONT_FAMILY } from '../common/Text';
import LoadingSpinner from '../common/LoadingSpinner';
import { PopupContext } from '../../context/Popup.context';
import { scale, scaleHeight } from '../../platformUtils';
import { APP_STATE, ANALYTICS_EVENTS } from '../../constants';
import { gameCardReloadStrings } from '../../stringConstants';
import { color } from '../../styles';
import { UserContext } from '../../context/User.context';
import { BackgroundMusicContext } from '../../context/BackgroundMusic.context';
import { logEvent } from '../../amplitudeUtils';

const Container = styled.View`
  height: 100%;
  width: 100%;
`;

const PromoCodeContainer = styled.View`
  margin-top: ${scaleHeight(20)};
  margin-bottom: ${scaleHeight(20)};
`;

const PromoCodeMessage = styled(Text)`
  color: ${color.white};
  margin-top: ${scale(18)};
  text-align: center;
`;

export const styles = StyleSheet.create({
  scrollViewContainerStyle: {
    paddingBottom: scaleHeight(16)
  }
});

const TokenPacksContainer = ({ isUserLoggedIn, navigation }) => {
  const [tokenPackData, setTokenPackData] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const { displayRequestError } = useContext(PopupContext);
  const { currentAppliedPromoCode, setCurrentAppliedPromoCode, shouldRefreshTokenPack, fetchPoints } = useContext(UserContext);
  const { appState } = useContext(BackgroundMusicContext);

  let isDisabled = false;


  const fetchAppStoreTokens = async tokenPacks => {
    const tokenIds = Platform.OS === 'android' ?
      tokenPacks.map(({ androidProductId }) => {
        if (androidProductId === "")
          return null
        return androidProductId
      }).filter(Boolean) :
      tokenPacks.map(({ iOSProductId }) => iOSProductId);
    try {
      const result = await RNIap.getProducts(tokenIds);
      const tokenPacksWithIapData = tokenPacks
        .map(token => {
          const iapData = result.find(({ productId }) => {
            if (Platform.OS === 'android') {
              return productId === token.androidProductId
            }
            return productId === token.iOSProductId;
          });
          return iapData ? { ...token, iapData } : {};
        })
        .filter(pack => pack.iapData);
      setTokenPackData(tokenPacksWithIapData);
    } catch (error) {
      displayRequestError(error.message);
    }
  };

  const fetchTokenPacks = async code => {
    setIsLoading(true);
    try {
      const res = await api.getTokenPacks(code, Platform.OS === 'android');
      if (res.status === 200 && res.data) {
        await fetchAppStoreTokens(res.data);
      }
    } catch (error) {
      displayRequestError(error.message);
    }
    setIsLoading(false);
  };

 

  useEffect(() => {
    // Check if deep link has promoCode meta data
    // checkBranchParams();
    fetchTokenPacks();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (shouldRefreshTokenPack) {
      // Token pack was successfully purchased - refresh
      fetchTokenPacks();
      fetchPoints();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [shouldRefreshTokenPack]);

  useEffect(() => {
    fetchTokenPacks(currentAppliedPromoCode);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentAppliedPromoCode]);

  // useEffect(() => {
  //   // Check branch params if app state became active from deep link
  //   if (appState === APP_STATE.ACTIVE) {
  //     // checkBranchParams();
  //   }
  //   // eslint-disable-next-line react-hooks/exhaustive-deps
  // }, [appState]);

  const requestPurchase = async productId => {
    try {
      await RNIap.initConnection()
      await RNIap.requestPurchase(productId, false);
      isDisabled = false;
    } catch (error) {
      isDisabled = false;
      // Error handled in purchase listener
    }
  };

  const onTokenPurchasePress = iOSProductId => {
    // if (!isUserLoggedIn) {
    //   setIsCreateAccountPopUpShown(true);
    // } else {
    requestPurchase(iOSProductId);
    // }
  };

  const handleIapPurchase = (productId) => {
    if(isDisabled) return;
      isDisabled = true;
      onTokenPurchasePress(productId)
  }

  /* eslint-disable react/prop-types */
  const renderTokenCard = ({
    imageUrl,
    name,
    tokenAmount,
    isOnSale,
    bonusTokenAmount,
    ribbonColor,
    ribbonName,
    backgroundColor,
    iapData,
    iOSProductId,
    androidProductId
  }) => (
    <TokenCard
      imageUrl={imageUrl}
      name={name}
      isOnSale={isOnSale}
      tokenAmount={tokenAmount}
      bonusTokenAmount={bonusTokenAmount}
      ribbonColor={ribbonColor}
      ribbonName={ribbonName}
      backgroundColor={backgroundColor}
      iapData={iapData}
      onPress={() => handleIapPurchase(Platform.OS === "android" ? androidProductId : iOSProductId)}
    />
  );
  /* eslint-enable react/prop-types */

 


  return (
    <Container>
      <KeyboardAwareScrollView
        extraScrollHeight={30}
      >
        <KeyboardAwareFlatList
          contentContainerStyle={styles.scrollViewContainerStyle}
          data={tokenPackData}
          renderItem={({ item }) => renderTokenCard(item)}
          keyExtractor={item => item.name}
          numColumns={1}
          // columnWrapperStyle={flatListStyles.columnStyle}
          initialNumToRender={4}
        />



      </KeyboardAwareScrollView>

      <LoadingSpinner isLoading={isLoading} />
    </Container>
  );
};

TokenPacksContainer.propTypes = {
  isUserLoggedIn: PropTypes.bool.isRequired,
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired,

  }).isRequired
};

export default TokenPacksContainer;
