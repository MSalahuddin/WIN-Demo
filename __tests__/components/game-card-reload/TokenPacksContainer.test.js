import React from 'react';
import { render } from 'react-native-testing-library';

import MockProvider from '../../MockProvider';
import TokenPacksContainer from '../../../js/components/game-card-reload/TokenPacksContainer';



describe('Component TokenPacksContainer', () => {
  test('renders correctly when isUserLoggedIn', () => {
    const output = render(
      <MockProvider isUserLoggedIn>
        <TokenPacksContainer isUserLoggedIn />
      </MockProvider>
    );
    expect(output).toMatchSnapshot();
  });

  test('renders correctly when isUserLoggedIn={false}', () => {
    const output = render(
      <MockProvider>
        <TokenPacksContainer isUserLoggedIn={false} />
      </MockProvider>
    );
    expect(output).toMatchSnapshot();
  });
});
