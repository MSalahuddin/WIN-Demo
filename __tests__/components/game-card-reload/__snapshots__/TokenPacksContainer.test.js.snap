// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component TokenPacksContainer renders correctly when isUserLoggedIn 1`] = `
<View
  style={
    Array [
      Object {
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={30}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <RCTScrollView
        automaticallyAdjustContentInsets={false}
        contentContainerStyle={
          Object {
            "paddingBottom": 32,
          }
        }
        contentInset={
          Object {
            "bottom": 0,
          }
        }
        data={null}
        disableVirtualization={false}
        enableAutomaticScroll={true}
        enableOnAndroid={false}
        enableResetScrollToCoords={true}
        extraHeight={75}
        extraScrollHeight={0}
        getItem={[Function]}
        getItemCount={[Function]}
        getScrollResponder={[Function]}
        handleOnScroll={[Function]}
        horizontal={false}
        initialNumToRender={4}
        keyExtractor={[Function]}
        keyboardDismissMode="interactive"
        keyboardOpeningTime={250}
        keyboardSpace={0}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        resetKeyboardSpace={[Function]}
        scrollEventThrottle={1}
        scrollForExtraHeightOnAndroid={[Function]}
        scrollIntoView={[Function]}
        scrollToEnd={[Function]}
        scrollToFocusedInput={[Function]}
        scrollToPosition={[Function]}
        showsVerticalScrollIndicator={true}
        stickyHeaderIndices={Array []}
        update={[Function]}
        updateCellsBatchingPeriod={50}
        viewIsInsideTabBar={false}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <ActivityIndicator
      animating={true}
      color="#FFFFFF"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;

exports[`Component TokenPacksContainer renders correctly when isUserLoggedIn={false} 1`] = `
<View
  style={
    Array [
      Object {
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={30}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <RCTScrollView
        automaticallyAdjustContentInsets={false}
        contentContainerStyle={
          Object {
            "paddingBottom": 32,
          }
        }
        contentInset={
          Object {
            "bottom": 0,
          }
        }
        data={null}
        disableVirtualization={false}
        enableAutomaticScroll={true}
        enableOnAndroid={false}
        enableResetScrollToCoords={true}
        extraHeight={75}
        extraScrollHeight={0}
        getItem={[Function]}
        getItemCount={[Function]}
        getScrollResponder={[Function]}
        handleOnScroll={[Function]}
        horizontal={false}
        initialNumToRender={4}
        keyExtractor={[Function]}
        keyboardDismissMode="interactive"
        keyboardOpeningTime={250}
        keyboardSpace={0}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        resetKeyboardSpace={[Function]}
        scrollEventThrottle={1}
        scrollForExtraHeightOnAndroid={[Function]}
        scrollIntoView={[Function]}
        scrollToEnd={[Function]}
        scrollToFocusedInput={[Function]}
        scrollToPosition={[Function]}
        showsVerticalScrollIndicator={true}
        stickyHeaderIndices={Array []}
        update={[Function]}
        updateCellsBatchingPeriod={50}
        viewIsInsideTabBar={false}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <ActivityIndicator
      animating={true}
      color="#FFFFFF"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;
